flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/nonha/embedded/EmbeddedLeaderService.java

```diff
@@ -506,33 +515,28 @@ public void run() {
 
 	private static class GrantLeadershipCall implements Runnable {
 
-		private final EmbeddedLeaderElectionService leaderElectionService;
+		private final LeaderContender contender;
 		private final UUID leaderSessionId;
 		private final Logger logger;
 
 		GrantLeadershipCall(
-				EmbeddedLeaderElectionService leaderElectionService,
+				LeaderContender contender,
 				UUID leaderSessionId,
 				Logger logger) {
 
-			this.leaderElectionService = checkNotNull(leaderElectionService);
+			this.contender = checkNotNull(contender);
 			this.leaderSessionId = checkNotNull(leaderSessionId);
 			this.logger = checkNotNull(logger);
 		}
 
 		@Override
 		public void run() {
-			leaderElectionService.isLeader = true;
-
-			final LeaderContender contender = leaderElectionService.contender;
-
 			try {
 				contender.grantLeadership(leaderSessionId);
 			}
 			catch (Throwable t) {
 				logger.warn("Error granting leadership to contender", t);
 				contender.handleError(t instanceof Exception ? (Exception) t : new Exception(t));
-				leaderElectionService.isLeader = false;
```

Comment:
The flag is no longer reset if the grant fails.

I'm wondering whether we need this flag at all. It's only being used in `EmbeddedLeaderElectionService#hasLeadership`, but shouldn't the `leaderSessionId` be sufficient for checking that? This seems especially true now that `updateLeader` updates both `isLeader` and `currentLeaderSessionId` right after another.

Reply:
I kept the flag because if we would remove it, then a user of the `EmbeddedLeaderElectionService#hasLeadership` could get a false positive if it guessed the current leadership id by chance.

It is true that we don't reset the `isLeader` flat to false in case of a failure. The assumption is that the `leaderContender` will stop the leader election service in case of a failure.